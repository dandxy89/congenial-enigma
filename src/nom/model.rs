#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ComparisonOp {
    LT,
    LTE,
    EQ,
    GT,
    GTE,
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Sense {
    Minimize,
    Maximize,
}

impl Sense {
    pub fn is_minimization(&self) -> bool {
        self == &Sense::Minimize
    }
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, PartialEq, Eq)]
pub enum SOSType {
    S1,
    S2,
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, PartialEq)]
pub enum Bound<'a> {
    Free(&'a str),
    LowerBound(&'a str, f64),
    UpperBound(&'a str, f64),
    DoubleBound(&'a str, f64, f64),
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, PartialEq)]
pub struct Coefficient<'a> {
    pub var_name: &'a str,
    pub coefficient: f64,
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[cfg_attr(feature = "serde", serde(tag = "type"))]
#[derive(Debug, PartialEq)]
pub enum Constraint<'a> {
    Standard { name: Option<String>, coefficients: Vec<Coefficient<'a>>, operator: ComparisonOp, rhs: f64 },
    SOS { name: Option<String>, sos_type: SOSType, weights: Vec<Coefficient<'a>> },
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[derive(Debug, PartialEq)]
pub struct Objective<'a> {
    pub name: Option<String>,
    pub coefficients: Vec<Coefficient<'a>>,
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, PartialEq)]
pub enum VariableType {
    Free,
    Binary,
    Integer,
    General { lower_bound: Option<f64>, upper_bound: Option<f64> },
    SemiContinuous { lower_bound: Option<f64>, upper_bound: Option<f64> },
}

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, PartialEq)]
pub struct Variable<'a> {
    pub name: &'a str,
    pub var_type: VariableType,
}

#[cfg(feature = "serde")]
impl<'de: 'a, 'a> serde::Deserialize<'de> for Constraint<'a> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        // Define the fields we expect to see
        #[derive(PartialEq, serde::Deserialize)]
        #[serde(field_identifier, rename_all = "lowercase")]
        enum Field {
            Type,
            Name,
            Coefficients,
            Weights,
            Operator,
            Rhs,
            SosType,
        }

        struct ConstraintVisitor<'a>(std::marker::PhantomData<Constraint<'a>>);

        impl<'de: 'a, 'a> serde::de::Visitor<'de> for ConstraintVisitor<'a> {
            type Value = Constraint<'a>;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("struct Constraint")
            }

            fn visit_map<V>(self, mut map: V) -> Result<Constraint<'a>, V::Error>
            where
                V: serde::de::MapAccess<'de>,
            {
                let constraint_type: String = match map.next_key::<Field>()? {
                    Some(Field::Type) => map.next_value()?,
                    _ => return Err(serde::de::Error::missing_field("type")),
                };
                match constraint_type.as_str() {
                    "Standard" => {
                        let mut name = None;
                        let mut coefficients = None;
                        let mut operator = None;
                        let mut rhs = None;

                        while let Some(key) = map.next_key()? {
                            match key {
                                Field::Name => name = Some(map.next_value()?),
                                Field::Coefficients => coefficients = Some(map.next_value()?),
                                Field::Operator => operator = Some(map.next_value()?),
                                Field::Rhs => rhs = Some(map.next_value()?),
                                _ => {
                                    let _ = map.next_value::<serde::de::IgnoredAny>()?;
                                }
                            }
                        }

                        Ok(Constraint::Standard {
                            name,
                            coefficients: coefficients.ok_or_else(|| serde::de::Error::missing_field("coefficients"))?,
                            operator: operator.ok_or_else(|| serde::de::Error::missing_field("operator"))?,
                            rhs: rhs.ok_or_else(|| serde::de::Error::missing_field("rhs"))?,
                        })
                    }
                    "SOS" => {
                        let mut name = None;
                        let mut sos_type = None;
                        let mut weights = None;

                        while let Some(key) = map.next_key()? {
                            match key {
                                Field::Name => name = Some(map.next_value()?),
                                Field::SosType => sos_type = Some(map.next_value()?),
                                Field::Weights => weights = Some(map.next_value()?),
                                _ => {
                                    let _ = map.next_value::<serde::de::IgnoredAny>()?;
                                }
                            }
                        }

                        Ok(Constraint::SOS {
                            name,
                            sos_type: sos_type.ok_or_else(|| serde::de::Error::missing_field("sos_type"))?,
                            weights: weights.ok_or_else(|| serde::de::Error::missing_field("weights"))?,
                        })
                    }
                    _ => Err(serde::de::Error::unknown_variant(&constraint_type, &["Standard", "SOS"])),
                }
            }
        }

        const FIELDS: &[&str] = &["type", "name", "coefficients", "weights", "operator", "rhs", "sos_type"];
        deserializer.deserialize_struct("Constraint", FIELDS, ConstraintVisitor(std::marker::PhantomData))
    }
}

#[cfg(feature = "serde")]
impl<'de: 'a, 'a> serde::Deserialize<'de> for Objective<'a> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        #[derive(serde::Deserialize)]
        #[serde(field_identifier, rename_all = "lowercase")]
        enum Field {
            Name,
            Coefficients,
        }

        struct ObjectiveVisitor<'a>(std::marker::PhantomData<Objective<'a>>);

        impl<'de: 'a, 'a> serde::de::Visitor<'de> for ObjectiveVisitor<'a> {
            type Value = Objective<'a>;

            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
                formatter.write_str("struct Objective")
            }

            fn visit_map<V>(self, mut map: V) -> Result<Objective<'a>, V::Error>
            where
                V: serde::de::MapAccess<'de>,
            {
                let mut name = None;
                let mut coefficients = None;

                while let Some(key) = map.next_key()? {
                    match key {
                        Field::Name => name = Some(map.next_value()?),
                        Field::Coefficients => coefficients = Some(map.next_value()?),
                    }
                }

                Ok(Objective { name, coefficients: coefficients.ok_or_else(|| serde::de::Error::missing_field("coefficients"))? })
            }
        }

        deserializer.deserialize_struct("Objective", &["name", "coefficients"], ObjectiveVisitor(std::marker::PhantomData))
    }
}
