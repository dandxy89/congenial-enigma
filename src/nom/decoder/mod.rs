pub mod coefficient;
pub mod constraint;
pub mod number;
pub mod objective;
pub mod problem_name;
pub mod sense;
pub mod sos_constraint;
pub mod variable;
