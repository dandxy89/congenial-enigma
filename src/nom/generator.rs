use std::sync::atomic::{AtomicU64, Ordering};

#[derive(Debug)]
pub struct IdGenerator {
    counter: AtomicU64,
    prefix: &'static str,
}

impl IdGenerator {
    pub fn new(prefix: &'static str) -> Self {
        Self { counter: AtomicU64::new(0), prefix }
    }

    pub fn next_id(&self) -> String {
        let id = self.counter.fetch_add(1, Ordering::Relaxed);
        format!("{}_{}", self.prefix, id)
    }
}

#[cfg(test)]
mod test {
    use crate::nom::generator::IdGenerator;

    #[test]
    fn test_id_generator() {
        let id_gen = IdGenerator::new("test");
        assert_eq!(id_gen.next_id(), "test_0");
        assert_eq!(id_gen.next_id(), "test_1");
    }
}
