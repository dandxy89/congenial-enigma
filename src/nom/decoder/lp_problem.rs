use nom::sequence::tuple;

use crate::nom::{
    decoder::{objective::objectives_section, problem_name::parse_comments, sense::parse_problem_sense},
    model::{Objective, Sense},
};

pub fn parse_lp_file(input: &str) -> Result<(Option<&str>, Sense, Vec<Objective>), nom::Err<nom::error::Error<&str>>> {
    // Parse the full structure with tuple combinator
    let (remaining, (comments, sense, objectives)) = tuple((
        // LP Problem Name
        parse_comments,
        // LP Sense
        parse_problem_sense,
        // Objective Section
        objectives_section,
    ))(input)?;

    // Make sure we consumed all input (or handle remaining if needed)
    if !remaining.trim().is_empty() {
        println!("Warning: Unparsed content remains: {remaining}");
    }

    Ok((comments, sense, objectives))
}

#[cfg(test)]
mod test {
    use crate::nom::{decoder::lp_problem::parse_lp_file, model::Sense};

    #[test]
    fn test_sections_so_far() {
        let input = "\\ This file has been generated by Author
\\ ENCODING=ISO-8859-1
\\Problem name: diet
Minimize
 obj1: -0.5 x - 2y - 8z
 obj2: y + x + z
 obj3: 10z - 2.5x
       + y";
        let parsed_result = parse_lp_file(input);
        assert!(parsed_result.is_ok());

        let (comments, sense, objectives) = parsed_result.unwrap();
        assert!(comments.is_some());
        assert!(matches!(sense, Sense::Minimize));
        insta::assert_debug_snapshot!(objectives);
    }
}
