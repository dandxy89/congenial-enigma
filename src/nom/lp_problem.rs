use std::{
    borrow::Cow,
    collections::{hash_map::Entry, HashMap},
};

use nom::{combinator::opt, sequence::tuple};

use crate::nom::{
    decoder::{
        constraint::{parse_constraint_header, parse_constraints},
        objective::parse_objectives,
        problem_name::parse_problem_name,
        sense::parse_sense,
        variable::{parse_binary_section, parse_bounds_section, parse_generals_section, parse_integer_section, parse_semi_section},
    },
    model::{Constraint, Objective, Sense, Variable},
    take_until_parser, ALL_VAR_BOUND_HEADERS, BINARY_HEADERS, BOUND_HEADERS, CONSTRAINT_HEADERS, GENERAL_HEADERS, SEMI_HEADERS,
    SOS_HEADERS,
};

#[cfg_attr(feature = "diff", derive(diff::Diff), diff(attr(#[derive(Debug, PartialEq)])))]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, Default, PartialEq)]
pub struct LPProblem<'a> {
    pub name: Option<&'a str>,
    pub sense: Sense,
    pub objectives: HashMap<&'a str, Objective<'a>>,
    pub constraints: HashMap<Cow<'a, str>, Constraint<'a>>,
    pub variables: HashMap<&'a str, Variable<'a>>,
}

impl LPProblem<'_> {
    #[inline]
    /// Returns the name of the LP Problem
    pub fn name(&self) -> Option<&str> {
        self.name
    }

    #[inline]
    /// Returns `true` if the `Self` a Minimize LP Problem
    pub fn is_minimization(&self) -> bool {
        self.sense.is_minimization()
    }

    #[inline]
    /// Returns the number of constraints contained within the Problem
    pub fn constraint_count(&self) -> usize {
        self.constraints.len()
    }

    #[inline]
    /// Returns the number of objectives contained within the Problem
    pub fn objective_count(&self) -> usize {
        self.objectives.len()
    }

    #[inline]
    /// Returns the number of variables contained within the Problem
    pub fn variable_count(&self) -> usize {
        self.variables.len()
    }
}

impl<'a> TryFrom<&'a str> for LPProblem<'a> {
    type Error = nom::Err<nom::error::Error<&'a str>>;

    fn try_from(input: &'a str) -> Result<Self, Self::Error> {
        let (input, (name, sense, obj_section, _cons_header)) =
            tuple((parse_problem_name, parse_sense, take_until_parser(&CONSTRAINT_HEADERS), parse_constraint_header))(input)?;
        let (_, (objs, mut variables)) = parse_objectives(obj_section)?;

        let (input, constraint_str) = take_until_parser(&BOUND_HEADERS)(input)?;
        let (_, (constraints, constraint_vars)) = parse_constraints(constraint_str)?;
        variables.extend(constraint_vars);

        let (input, bound_str) = take_until_parser(&ALL_VAR_BOUND_HEADERS)(input)?;
        let (_, bounds) = parse_bounds_section(bound_str)?;

        for (bound_name, var_type) in bounds {
            match variables.entry(bound_name) {
                Entry::Occupied(mut occupied_entry) => {
                    occupied_entry.get_mut().set_vt(var_type);
                }
                Entry::Vacant(vacant_entry) => {
                    vacant_entry.insert(Variable { name: bound_name, var_type });
                }
            }
        }

        let (input, integer_str) = opt(take_until_parser(&GENERAL_HEADERS))(input)?;
        if let Some(integer_str) = integer_str {
            let (_, _integer_var) = parse_integer_section(integer_str)?;
        }

        let (input, generals_str) = opt(take_until_parser(&BINARY_HEADERS))(input)?;
        if let Some(generals_str) = generals_str {
            let (_, _general_var) = parse_generals_section(generals_str)?;
        }

        let (_input, binary_str) = opt(take_until_parser(&SEMI_HEADERS))(input)?;
        if let Some(binary_str) = binary_str {
            let (_, _binary_vars) = parse_binary_section(binary_str)?;
        }

        let (_input, semi_str) = opt(take_until_parser(&SOS_HEADERS))(input)?;
        if let Some(semi_str) = semi_str {
            let (_, _semi_continuous_vars) = parse_semi_section(semi_str)?;
        }

        Ok(LPProblem { name, sense, objectives: objs, constraints, variables })
    }
}

#[cfg(test)]
mod test {
    const SMALL_INPUT: &str = "\\ This file has been generated by Author
\\ ENCODING=ISO-8859-1
\\Problem name: diet
Minimize
 obj1: -0.5 x - 2y - 8z
 obj2: y + x + z
 obj3: 10z - 2.5x
       + y
subject to:
c1:  3 x1 + x2 + 2 x3 = 30
c2:  2 x1 + x2 + 3 x3 + x4 >= 15
c3:  2 x2 + 3 x4 <= 25
bounds
x1 free
x2 >= 1
100 <= x2dfsdf <= -1
Generals
Route_A_1
Route_A_2
Route_A_3
End";

    #[cfg(feature = "serde")]
    #[test]
    fn test_minified_example() {
        let problem = crate::nom::lp_problem::LPProblem::try_from(SMALL_INPUT).expect("test case not to fail");

        assert_eq!(problem.objectives.len(), 3);
        assert_eq!(problem.constraints.len(), 3);

        insta::assert_yaml_snapshot!(&problem, {
            ".objectives" => insta::sorted_redaction(),
            ".constraints" => insta::sorted_redaction(),
            ".variables" => insta::sorted_redaction()
        });
    }

    #[cfg(feature = "serde")]
    #[test]
    fn test_serialization_lifecycle() {
        let problem = crate::nom::lp_problem::LPProblem::try_from(SMALL_INPUT).expect("test case not to fail");
        // Serialised
        let serialized_problem = serde_json::to_string(&problem).expect("test case not to fail");
        // Deserialise
        let _: crate::nom::lp_problem::LPProblem<'_> = serde_json::from_str(&serialized_problem).expect("test case not to fail");
    }
}
