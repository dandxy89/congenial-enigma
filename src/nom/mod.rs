pub mod decoder;
pub mod generator;
pub mod lp_problem;
pub mod model;

use nom::{error::ErrorKind, IResult};

pub const CONSTRAINT_HEADERS: [&str; 4] = ["subject to", "such that", "s.t.", "st:"];

pub const BINARY_HEADERS: [&str; 4] = ["binaries", "binary", "bin", "end"];
pub const ALL_BOUND_HEADERS: [&str; 13] = [
    "bounds",
    "bound",
    "generals",
    "general",
    "integers",
    "integer",
    "binaries",
    "binary",
    "bin",
    "semi-continuous",
    "semis",
    "semi",
    "end",
];
pub const GENERAL_HEADERS: [&str; 3] = ["generals", "general", "end"];
pub const INTEGER_HEADERS: [&str; 3] = ["integers", "integer", "end"];
pub const SEMI_HEADERS: [&str; 4] = ["semi-continuous", "semis", "semi", "end"];
pub const SOS_HEADERS: [&str; 2] = ["sos", "end"];

pub const VALID_LP_CHARS: [char; 18] = ['!', '#', '$', '%', '&', '(', ')', '_', ',', '.', ';', '?', '@', '\\', '{', '}', '~', '\''];

pub fn log_remaining(prefix: &str, remaining: &str) {
    if !remaining.trim().is_empty() {
        log::debug!("{prefix}: {remaining}");
        println!("{prefix}: {remaining}"); // Remove once branch is complete
    }
}

fn take_until_cased<'a>(tag: &'a str) -> impl Fn(&'a str) -> IResult<&'a str, &'a str> {
    move |input: &str| {
        let mut index = 0;
        let tag_lower = tag.to_lowercase();
        let chars: Vec<char> = input.chars().collect();

        if chars.len() < tag.len() {
            return Err(nom::Err::Error(nom::error::Error::new(input, ErrorKind::TakeUntil)));
        }

        while index <= chars.len() - tag.len() {
            let window: String = chars[index..index + tag.len()].iter().collect();
            if window.to_lowercase() == tag_lower {
                return Ok((&input[index..], &input[..index]));
            }
            index += 1;
        }

        Err(nom::Err::Error(nom::error::Error::new(input, ErrorKind::TakeUntil)))
    }
}

pub fn take_until_parser<'a>(tags: &'a [&'a str]) -> impl Fn(&'a str) -> IResult<&'a str, &'a str> + 'a {
    move |input| {
        tags.iter().map(|&tag| take_until_cased(tag)).fold(
            Err(nom::Err::Error(nom::error::Error::new(input, nom::error::ErrorKind::Alt))),
            |acc, parser| match acc {
                Ok(ok) => Ok(ok),
                Err(_) => parser(input),
            },
        )
    }
}
