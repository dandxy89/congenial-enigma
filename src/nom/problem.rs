use crate::nom::model::{Constraint, Objective, Sense, Variable};

#[derive(Debug, PartialEq)]
pub struct LPProblem<'a> {
    pub name: Option<&'a str>,
    pub sense: Sense,
    pub objectives: Vec<Objective<'a>>,
    pub constraints: Vec<Constraint<'a>>,
    pub variables: Vec<Variable<'a>>,
}

impl<'a> LPProblem<'a> {
    #[inline]
    pub fn name(&self) -> Option<&str> {
        self.name
    }

    #[inline]
    pub fn is_minimization(&self) -> bool {
        self.sense.is_minimization()
    }

    #[inline]
    pub fn constraint_count(&self) -> usize {
        self.constraints.len()
    }

    #[inline]
    pub fn objective_count(&self) -> usize {
        self.objectives.len()
    }
}
