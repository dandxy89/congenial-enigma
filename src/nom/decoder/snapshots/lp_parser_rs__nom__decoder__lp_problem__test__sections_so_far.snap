---
source: src/nom/decoder/lp_problem.rs
expression: objectives
---
[
    Objective {
        name: Some(
            "obj1",
        ),
        coefficients: [
            Coefficient {
                var_name: "x",
                coefficient: -0.5,
            },
            Coefficient {
                var_name: "y",
                coefficient: -2.0,
            },
            Coefficient {
                var_name: "z",
                coefficient: -8.0,
            },
        ],
    },
    Objective {
        name: Some(
            "obj2",
        ),
        coefficients: [
            Coefficient {
                var_name: "y",
                coefficient: 1.0,
            },
            Coefficient {
                var_name: "x",
                coefficient: 1.0,
            },
            Coefficient {
                var_name: "z",
                coefficient: 1.0,
            },
        ],
    },
    Objective {
        name: Some(
            "obj3",
        ),
        coefficients: [
            Coefficient {
                var_name: "z",
                coefficient: 10.0,
            },
            Coefficient {
                var_name: "x",
                coefficient: -2.5,
            },
            Coefficient {
                var_name: "y",
                coefficient: 1.0,
            },
        ],
    },
]
